### ISSUE TITLE
React Context causes infinite re-render loop in page component marked as async

### ISSUE THREAD

**Issue by JGH153** [vercel/next.js#51528]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Pro
Binaries:
  Node: 18.16.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant packages:
  next: 13.4.7-canary.1
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 4.9.5
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/JGH153/next-13-app-context-bug

### To Reproduce

1. Clone project
2. npm install & npm run dev
3. Open http://localhost:3000/
4. Click on URL to about page
5. Click on button

### Describe the Bug

Using context in page components causes infinite re-renders when updating context value if marked as async.

This works:
`export default function About() {`

This causes infinite re-renders on context update:
`export default async function About() {`

Both are maked as `"use client";`

### Expected Behavior

Should not cause the tab to freeze in an infinite re-render loop, or provide a warning/error that "use client" and async should not be used together in the same page component with context

**Comment by DuCanhGH** [comment]
Client Components can't be marked with `async`. Perhaps an error should be thrown, but yeah.

What you can do is extract parts that need to use the context out of the page, and do whatever asynchronous inside the page, then pass the result to the extracted (client) components should they need it.

**Comment by JGH153** [comment]
I got this issue as I refactored the component from a server component to a client component, but forgot to remove `async`.

**Comment by jokull** [comment]
Spent a day on this before I saw that my client component was marked async. There are no lint errors, warnings or anything. I highly suggest throwing an error or doing some static checking of this in the bundler or dev server. It will help a lot of people. 🙏🏻

**Comment by markhc** [comment]
I can also attest for the time spent trying to figure out what the hell was going wrong. Turns out my component cannot be made async, which I didn't know until I found this thread.

This really should be made more obvious, be it with runtime or linter errors.

**Comment by JohnGemstone** [comment]
Yep this had me scratching my head for a bit, async client components crash my Chrome browser whenever theres a hotreload change, no warnings / errors. Using the latest 13.4.8

**Comment by nevenhsu** [comment]
Same here! 
After changing the component from server to client, there is no warning to notify me this will cause infinite loop.

**Comment by AyanCodes** [comment]
Hi! i'm having the same problem
Does anyone have a workaround to this?
```
async function getCurrentUser() {
  return new Promise((resolve, reject) => {
     const unsubscribe = onAuthStateChanged(auth, (user) => {
        unsubscribe();
        console.log("User")
        resolve(user);
     }, reject);
  });
}

export default async function Page() {
  const user = await getCurrentUser()
  console.log(user)
  return ....
  }
```

**Comment by MattOfficial** [comment]
> Hi! i'm having the same problem with firebase waiting for user, while this code works it causes infinite loop and this code doesn't seem to work on "use server" either Does anyone have a workaround to this?
> 
> ```
> async function getCurrentUser() {
>   return new Promise((resolve, reject) => {
>      const unsubscribe = onAuthStateChanged(auth, (user) => {
>         unsubscribe();
>         console.log("User")
>         resolve(user);
>      }, reject);
>   });
> }
> 
> export default async function Page() {
>   const user = await getCurrentUser()
>   console.log(user)
>   return ....
>   }
> ```

Maybe you can try something like this:
`
'use client';
import { useEffect } from React;
// Other imports

async function getCurrentUser() {
  return new Promise((resolve, reject) => {
     const unsubscribe = onAuthStateChanged(auth, (user) => {
        unsubscribe();
        console.log("User")
        resolve(user);
     }, reject);
  });
}

export default function Page() {

useEffect(() => {
// using IIFE. But you can define a function and call it inside useEffect
    (async function() {
      const user = await getCurrentUser()
      console.log(user)
    })();
}, [])
  
  return ....
  }
`

**Comment by Patrick-Tomic** [comment]
I had this issue when I was creating my first full-stack project with next, it took me a minute to see the error so I found this one quick.

When it comes to contributing to open source projects, do I just look at the error and find the fix or is there a way to commit the work?

**Comment by marcoshernanz** [comment]
Thanks for the repro. This is the same unsupported async Client Component pattern covered by #50898. #51547 added the `@next/next/no-async-client-component` warning in 13.4.8, and current Next.js/React now surfaces a runtime error instead of hanging. I verified the repro against current stable and canary, and both show the error page rather than entering an infinite render loop. Closing as resolved.
